pkg/event/event.go

```diff
@@ -0,0 +1,441 @@
+package event
+
+import (
+	"encoding/json"
+	"fmt"
+	"maps"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/argocd"
+	"github.com/akuity/kargo/pkg/expressions"
+	"github.com/akuity/kargo/pkg/x/promotion/runner/builtin"
+)
+
+// A list of all known event types
+var KnownEventTypes = []kargoapi.EventType{
+	// NOTE(thomastaylor312): This list should be updated if new event types are added
+	kargoapi.EventTypePromotionCreated,
+	kargoapi.EventTypePromotionSucceeded,
+	kargoapi.EventTypePromotionFailed,
+	kargoapi.EventTypePromotionErrored,
+	kargoapi.EventTypePromotionAborted,
+	kargoapi.EventTypeFreightApproved,
+	kargoapi.EventTypeFreightVerificationSucceeded,
+	kargoapi.EventTypeFreightVerificationFailed,
+	kargoapi.EventTypeFreightVerificationErrored,
+	kargoapi.EventTypeFreightVerificationAborted,
+	kargoapi.EventTypeFreightVerificationInconclusive,
+	kargoapi.EventTypeFreightVerificationUnknown,
+}
+
+// Meta is an interface for our built in event types that allows for easier conversion to
+// CloudEvent types
+type Meta interface {
+	// Type returns the event type for the struct
+	Type() kargoapi.EventType
+	// Kind returns the object kind the event is related to (e.g. "Promotion", "Freight", etc.).
+	// This is used for constructing a valid object reference for the event source
+	Kind() string
+	// GetProject returns the project associated with the event
+	GetProject() string
+	// GetName returns the name of the object associated with the event
+	GetName() string
+}
+
+// Message is an interface for setting and getting the message of any built in event
+type Message interface {
+	// GetMessage returns the message of the event
+	GetMessage() string
+	// SetMessage sets the message of the event
+	SetMessage(string)
+}
+
+// AnnotationMarshaler is an interface for any type that can marshal out to our custom annotation
+// types
+type AnnotationMarshaler interface {
+	MarshalAnnotations() map[string]string
+}
+
+// Common is a struct that contains fields common to all events.
+type Common struct {
+	Project string  `json:"project"`
+	Actor   *string `json:"actor,omitempty"`
+	Message string  `json:"message"`
+}
+
+func (c Common) GetProject() string {
+	return c.Project
+}
+
+func (c Common) GetMessage() string {
+	return c.Message
+}
+
+func (c *Common) SetMessage(message string) {
+	c.Message = message
+}
+
+// Promotion is a struct that contains common fields for promotion-related events.
+type Promotion struct {
+	*Freight
+	Name         string                 `json:"promotionName"`
```

Comment:
No strong opinion, but I notice it's `promotionName` over the wire. I know it stutters, but naming it `PromotionName` here might avoid confusion elsewhere in the code. I could see this being mistaken for the name of the _promotion event_. But as I said... I don't feel strongly.

Reply:
I was trying to avoid stuttering because I definitely think of this as the name of the promotion. Events usually are referred to in terms of Types or IDs and not names, so I lean toward keeping it as is